examples-next/auth/schema.ts

```diff
@@ -0,0 +1,144 @@
+import { createSchema, list } from '@keystone-spike/keystone/schema';
+import { text, checkbox, password } from '@keystone-spike/fields';
+
+/**
+ * TODO: with the new session API, we're breaking keystone's current assumptions about the
+ * `authentication` argument provided to access control functions. This will be a breaking change
+ * for developers using Keystone.
+ *
+ * This needs to be implemented (replace `authentication` with `session`) before the code below
+ * will work.
+ *
+ * While we're here I'd like to propose some other changes to improve the ergonomics; see the types
+ * below for my proposal. Note that I'm specifically omitting the graphql context and any way of
+ * executing queries because I'd like to review that API as well, separately (and we can add it in
+ * later). This may be a bit contentious, but we need a new object anyway so I'm happy if in the
+ * first pass we just implement ({ session, item }) argument so the code below works and we can
+ * discuss the rest of the proposal as a group.
+ *
+ * Note that custom schema access control deviates significantly from the list and field access
+ * control arguments; we need to revisit that in the context of how graphql schema extensions work
+ * now. See https://www.keystonejs.com/api/access-control#custom-schema-access-control
+ *
+ * Also note, I'm not sure the reasoning behind why ListAccess seems to be resolved once per
+ * mutation for mutations that operate on multiple items, while FieldAccess seems to be resolved
+ * once per item (many times per mutation) for the same. At least, that's what I think is going on.
+ *
+ * This seems like a real foot-gun, and I'd prefer that they're implemented consistently, in favor
+ * of aligning on FieldAccessInput and resolving the function once per item being updated in the
+ * case of mutations that operate on multiple items.
+ *
+ * Or maybe my understanding is wrong? This is what's implied by the documentation (see
+ * https://www.keystonejs.com/api/access-control#field-level-access-control) but then if it's not
+ * the case I don't see how `existingItem` works in mutations that operate on multiple items.
+ *
+ * I also have questions about `originalInput` in the case of nested mutations (e.g creating an
+ * item as the input to a relationship field). Is it the whole "original" input? Or just the nested
+ * input for the relationship? Seems like it would have to be the second, but it's not clear from
+ * the documentation. Is this consistent with how the `gqlName` argument works?
```

Comment:
`originalInput` is the entire input object as passed directly into the graphql mutation, regardless of what nested operation is running.

The point here is to give access to that input _before_ it's modified by defaultValues.

Reply:
Seems like you'd need to do an incredibly large amount of work to use that then.

If I'm checking input as part of resolving `create` access control for the user list, in this mutation:

```
mutation {
  createPost(data: {
    title: "Hello World"
    author: {
      create: {
        name: "Access Hack"
        isAdmin: true
      }
    }) {
    id
  }
}
```

If what I'm getting is the _entire_ input, I need to check for all the ways users can be created.

Or maybe I'm misreading it and the input would be what's passed to `author { create: { /* this input */ } }` in which case that makes a heap more sense 👍

Reply:
I _think_ it's the second (`author { create: { /* this input */ }`) 👍